pkg/model/components/cloudcontrollermanager.go

```diff
@@ -0,0 +1,123 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package components
+
+import (
+	"fmt"
+
+	//"github.com/golang/glog"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/apis/kops/util"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/loader"
+)
+
+type CloudControllerManagerOptionsBuilder struct {
+	Context *OptionsContext
+}
+
+var _ loader.OptionsBuilder = &CloudControllerManagerOptionsBuilder{}
+
+func (b *CloudControllerManagerOptionsBuilder) BuildOptions(o interface{}) error {
+
+	clusterSpec := o.(*kops.ClusterSpec)
+	if clusterSpec.CloudControllerManager == nil {
+		clusterSpec.CloudControllerManager = &kops.CloudControllerManagerConfig{}
+	}
+	ccm := clusterSpec.CloudControllerManager
+
+	if !clusterSpec.EnableExternalCloudController {
+		return nil
+	}
+
+	k8sv17, err := util.ParseKubernetesVersion("v1.7.0")
+	if err != nil {
+		return fmt.Errorf("Unable to parse kubernetes version %s", err)
+	}
+
+	kubernetesVersion, err := KubernetesVersion(clusterSpec)
+	if err != nil {
+		return fmt.Errorf("Unable to parse kubernetesVersion %s", err)
+	}
+
+	if !kubernetesVersion.GTE(*k8sv17) {
+		return fmt.Errorf("Kubernetes version should be >=1.7.0 to enable cloud controller manager")
+	}
+
+	switch kops.CloudProviderID(clusterSpec.CloudProvider) {
+	case kops.CloudProviderAWS:
+		ccm.CloudProvider = "aws"
+	case kops.CloudProviderGCE:
+		ccm.CloudProvider = "gce"
+	case kops.CloudProviderVSphere:
+		ccm.CloudProvider = "vsphere"
+	default:
+		return fmt.Errorf("unknown cloud provider %q", clusterSpec.CloudProvider)
+	}
+
+	ccm.LogLevel = 2
+
+	image, err := Image("cloud-controller-manager", clusterSpec, b.Context.AssetBuilder)
+	if err != nil {
+		return err
+	}
+	ccm.Image = image
+
+	// Doesn't seem to be any real downside to always doing a leader election
+	ccm.LeaderElection = &kops.LeaderElectionConfiguration{LeaderElect: fi.Bool(true)}
+
+	ccm.AllocateNodeCIDRs = fi.Bool(true)
+	ccm.ConfigureCloudRoutes = fi.Bool(false)
+
+	networking := clusterSpec.Networking
```

Comment:
So if we add another networking provider, they need to know about this code.  Can we move this code into https://github.com/kubernetes/kops/blob/8fb99a87e2eeac1f55ccaeb5c50ceb934f7cedd8/upup/pkg/fi/cloudup/networking.go?  We may have to refactor a bit, but we need to consolidate the code.